backtrace-library/src/main/java/backtraceio/library/breadcrumbs/BacktraceQueueFileHelper.java

```diff
@@ -0,0 +1,112 @@
+package backtraceio.library.breadcrumbs;
+
+import com.squareup.tape.QueueFile;
+
+import java.io.File;
+import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.nio.charset.StandardCharsets;
+
+import backtraceio.library.logger.BacktraceLogger;
+
+public class BacktraceQueueFileHelper {
+    /**
+     * The base directory of the breadcrumb logs
+     */
+    private String breadcrumbLogDirectory;
+
+    /**
+     * The breadcrumb storage file
+     */
+    private QueueFile breadcrumbStore;
+
+    private final String LOG_TAG = BacktraceQueueFileHelper.class.getSimpleName();
+
+    private int maxQueueFileSizeBytes;
+
+    // This minimum file size comes from QueueFile::INITIAL_LENGTH
+    private final int minimumQueueFileSizeBytes = 4096;
+
+    private final Method usedBytes;
+
+    // Let our exceptions bubble all the way up to BacktraceBreadcrumbsLogger constructor
+    // We definitely cannot construct BacktraceBreadcrumbsLogger without an open file log
+    public BacktraceQueueFileHelper(String breadcrumbLogDirectory, int maxQueueFileSizeBytes) throws IOException, NoSuchMethodException {
+        this.breadcrumbLogDirectory = breadcrumbLogDirectory;
+        breadcrumbStore = new QueueFile(new File(this.breadcrumbLogDirectory));
+
+        // QueueFile pre-allocates a file of a certain size and fills with empty data,
+        // so normal File operations will not give us an accurate count of the bytes
+        // in the file.
+        // Therefore we expose the private method QueueFile::usedBytes
+        usedBytes = QueueFile.class.getDeclaredMethod("usedBytes");
+        usedBytes.setAccessible(true);
+
+        if (maxQueueFileSizeBytes < minimumQueueFileSizeBytes) {
+            this.maxQueueFileSizeBytes = minimumQueueFileSizeBytes;
+        } else {
+            this.maxQueueFileSizeBytes = maxQueueFileSizeBytes;
+        }
+
+        Runtime.getRuntime().addShutdownHook(
+            new Thread() {
+                public void run() {
+                    shutdownHook();
+                }
+            }
+        );
+    }
+
+    public boolean add(byte[] bytes) {
+        try {
+            int usedBytes = (int) this.usedBytes.invoke(breadcrumbStore);
+            int breadcrumbLength = bytes.length;
+
+            if (breadcrumbLength > 4096) {
+                BacktraceLogger.e(LOG_TAG, "We should not have a breadcrumb this big, this is a bug!");
+                return false;
+            }
+
+            // We clear the space we need from the QueueFile first to prevent
+            // the QueueFile from expanding to accommodate the new breadcrumb
+            while (!breadcrumbStore.isEmpty() && (usedBytes + breadcrumbLength) > maxQueueFileSizeBytes) {
+                breadcrumbStore.remove();
+                usedBytes--;
```

Comment:
are you sure we want to remove only 1 byte from "usedBytes" variable?

Reply:
Thank you for catching this!